### PR TITLE
Cache Block Location to save memory

### PR DESCRIPTION
**Pull request by elega**
### What changes are proposed in this pull request?

Cache block location object to reduce memory consumption

### Why are the changes needed?

Number of possible Block location value is limited, up to 9 * number of workers. By caching it, memory consumption can be saved.

### Does this PR introduce any user facing changes?

N/A

### CODE REVIEW

**Comment by maobaolong** [comment]
@elega Is this the improvement of https://github.com/Alluxio/alluxio/pull/16763 ? As this is better than my PR, Shall I close #16763 ?

**Comment by elega** [comment]
@maobaolong yup thanks

File: core/common/src/main/java/alluxio/util/proto/BlockLocationUtils.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+   * Get a shared grpc block location object. If it does not exist, create and cache it.
+   * Because the valid values of tierAlias and mediumType are only MEM, SSD and HDD,
+   * The size of the cache map is limited.
+   *
+   * @param blockLocation the block location to cache
+   * @return a shared block location object from the cache
+   */
+  public static BlockLocation getCached(BlockLocation blockLocation) {
+    Preconditions.checkState(VALID_MEDIUM_TYPE_VALUES.contains(blockLocation.getTier()),
+        "TierAlias must be one of the following MEM, HDD and SSD but got %s",
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/util/proto/BlockLocationUtils.java:71]:
```suggestion
        "TierAlias must be one of {MEM, HDD and SSD} but got %s",
```

**Review by jiacheliu3 — APPROVED** [review]
LGTM with a minor comment, thanks!

File: core/common/src/main/java/alluxio/util/proto/BlockLocationUtils.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+   * The size of the cache map is limited.
+   *
+   * @param blockLocation the block location to cache
+   * @return a shared block location object from the cache
+   */
+  public static BlockLocation getCached(BlockLocation blockLocation) {
+    Preconditions.checkState(VALID_MEDIUM_TYPE_VALUES.contains(blockLocation.getTier()),
+        "TierAlias must be one of {MEM, HDD and SSD} but got %s",
+        blockLocation.getTier());
+    Preconditions.checkState(VALID_MEDIUM_TYPE_VALUES.contains(blockLocation.getMediumType()),
```
Comment (maobaolong) [inline core/common/src/main/java/alluxio/util/proto/BlockLocationUtils.java:73]: The supported mediumTypes are read from configuration too.
``` 
  PropertyKey tierDirMediumConf =
        PropertyKey.Template.WORKER_TIERED_STORE_LEVEL_DIRS_MEDIUMTYPE.format(mTierOrdinal);
    List<String> dirMedium = Configuration.getList(tierDirMediumConf);

```

**Review by maobaolong — COMMENTED** [review]
This PR can help to reduce heap cost extremely, but there are hard code about MediumType and tier alias.

File: core/common/src/main/java/alluxio/util/proto/BlockLocationUtils.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+ * An util class to create cached grpc block locations.
+ */
+public class BlockLocationUtils {
+  private static final IndexDefinition<BlockLocation, BlockLocation> OBJECT_INDEX =
+      IndexDefinition.ofUnique((b) -> b);
+
+  private static final IndexDefinition<BlockLocation, Long> WORKER_ID_INDEX =
+      IndexDefinition.ofNonUnique(BlockLocation::getWorkerId);
+
+  // TODO(maobaolong): Add a metric to monitor the size of mLocationCacheMap
```
[the remark this answers is not in the document]
Comment (elega) [reply core/common/src/main/java/alluxio/util/proto/BlockLocationUtils.java:33]: sure i can update this in the next PR

File: core/common/src/main/java/alluxio/util/proto/BlockLocationUtils.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  private static final IndexedSet<BlockLocation> BLOCK_LOCATION_CACHE =
+      new IndexedSet<>(OBJECT_INDEX, WORKER_ID_INDEX);
+
+  private static final Set<String> VALID_MEDIUM_TYPE_VALUES =
```
[the remark this answers is not in the document]
Comment (elega) [reply core/common/src/main/java/alluxio/util/proto/BlockLocationUtils.java:37]: i'll add a comment for this one

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**